media/Jsontracer/dynamicGraduation.js

```diff
@@ -0,0 +1,126 @@
+//    * Redistributions in binary form must reproduce the above
+// copyright notice, this list of conditions and the following disclaimer
+// in the documentation and/or other materials provided with the
+// distribution.
+//    * Neither the name of Google Inc. nor the names of its
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+// This file referenced the result of
+// https://github.com/catapult-project/catapult/tree/444aba89e1c30edf348c611a9df79e2376178ba8/tracing
+
+export default function dynamicGraduation() {
+  const ruler = document.querySelector('.ruler');
+
+  if (!ruler) {
+    return;
+  }
+
+  const graduation = document.querySelector('.graduation');
+  const graduationList = document.querySelectorAll('.graduation');
+  const body = document.querySelector('body');
+  const rulerBlank = document.querySelector('.ruler-blank');
+  const staticRulerWidth = body.clientWidth - rulerBlank.clientWidth;
+
+  const setData = document.querySelector('.set-data');
+  const timeLimit = setData.dataset['timeLimit'];
+  const digit = setData.dataset['digit'];
+  const initGraduationCnt = timeLimit / 10 ** (digit - 1);
+  const staticGraduationWidth = parseInt(staticRulerWidth / initGraduationCnt);
+
+  if (graduation.offsetWidth < staticGraduationWidth - 10) {
+    removeGraduation(ruler, graduationList.length);
+  } else if (graduation.offsetWidth < staticGraduationWidth * 2) {
+    return;
+  } else {
+    addGraduation(ruler, graduationList.length);
+  }
+  updateGraduation(ruler, timeLimit);
+}
+
+function removeGraduation(ruler, cnt) {
+  for (let i = 0; i < cnt / 2; i++) {
+    const graduation = document.querySelector('.graduation');
+    ruler.removeChild(graduation);
+  }
+}
+
+function addGraduation(ruler, cnt) {
+  for (let i = 0; i < cnt; i++) {
+    const tempGraduation = document.createElement('div');
+    tempGraduation.className = 'graduation';
+
+    for (let i = 0; i < 5; i++) {
+      const tempSmallGraduation = document.createElement('div');
+      tempSmallGraduation.className = 'small-graduation';
+
+      if (i === 0) {
```

Comment:
There are two `i` in for-loop. Please avoid variable name duplication :)

Reply:
thank you for your feedback, i will fix it